mixin-metrics.go

```diff
@@ -0,0 +1,319 @@
+			return err
+		}
+		*queries = append(*queries, v.(string))
+	}
+
+	return nil
+}
+
+// parses through a dashboard to extract queries, and then metrics from queries
+func ParseDashboard(file *os.File) (*MetricsFile, error) {
+
+	queries := make([]string, 0)
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+	res := map[string]interface{}{}
+
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	json.Unmarshal([]byte(bytes), &res)
+	if err := ParseJq(&queries, res, ".templating.list[].query"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".panels[]?.targets[]?.expr"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".rows[].panels[].targets[].expr"); err != nil {
+		return nil, err
+	}
+
+	for _, query := range queries {
+		if err := ParseQuery(query, metrics); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+	return &metricsFile, nil
+}
+
+// todo: separate rules and raw metrics
+// parses through a rules file and extracts queries, and then metrics from queries
+func ParseRules(file *os.File) (*MetricsFile, error) {
+
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+
+	var conf RuleConfig
+
+	rulesFile, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(rulesFile, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := conf.RuleGroups
+	for _, group := range groups {
+		for _, rule := range group.Rules {
+			err := ParseQuery(rule.Expr, metrics)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+
+	return &metricsFile, nil
+}
+
+// iterates over all rules/dash files in dir
+func ParseDir(dir string, isRules bool) (*MetricsDir, error) {
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	metricsDir := MetricsDir{
+		MetricsFiles: make([]MetricsFile, 0),
+	}
+
+	for _, fileInfo := range files {
+		fmt.Println("Parsing: ", fileInfo.Name())
+		file, err := os.Open(dir + "/" + fileInfo.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		if isRules {
+			metricsFile, err := ParseRules(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		} else {
+			metricsFile, err := ParseDashboard(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		}
+
+		err = file.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &metricsDir, nil
+
+}
+
+// write out MetricsDir to outputFile
+func (md *MetricsDir) WriteOut(outputFile string) error {
+	out, err := json.MarshalIndent(*md, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(outputFile, out, os.FileMode(int(0666))); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
since the structure of your code is essentially:

```
if err := ...; err != nil {
   return err
}

return nil
```

you can simplify this to:

```suggestion
	return ioutil.WriteFile(outputFile, out, os.FileMode(int(0666)))
```